src/cache/cache.py

```diff
@@ -0,0 +1,301 @@
+
+        Args:
+            address: Memory address to check
+
+        Returns:
+            Optional[CacheLine]: Copy of the cache line if present, None if not cached
+        """
+        addr_fields = self.__decompose_address(address)
+
+        # Return None if set doesn't exist yet
+        if self.sets[addr_fields.index] is None:
+            return None  # No need to allocate yet, not needed by this Cache
+
+        cache_set = self.sets[addr_fields.index]
+        way_index = cache_set.find_way_by_tag(addr_fields.tag)
+
+        if way_index is not None:
+            # Return a copy of the cache line to prevent accidental modifications
+            line = cache_set.ways[way_index]
+            return CacheLine(tag=line.tag, mesi_state=line.mesi_state, in_l1=line.in_l1)
+        return None
+
+    def __create_set(self) -> CacheSetPLRUMESI:
+        """
+        Internal method to create a new cache set dynamically
+        """
+        return CacheSetPLRUMESI(num_ways=self.associativity)
+
+    def get_line_state(self, address: int) -> Optional[MESIState]:
+        """
+        Get the MESI state of the cache line containing the address
+
+        Args:
+            address: Memory address to check
+        Returns:
+            MESIState if line exists, None if line not in cache
+        """
+        addr_fields = self.__decompose_address(address)
+
+        # Return None if set doesn't exist yet
+        if self.sets[addr_fields.index] is None:
+            return None
+
+        cache_set = self.sets[addr_fields.index]
+        way_index = cache_set.find_way_by_tag(addr_fields.tag)
+
+        if way_index is not None:
+            return cache_set.mesi_state[way_index]
+        return None
+
+    def set_line_state(self, address: int, state: MESIState) -> bool:
+        """
+        Setter method to modify the MESI state of a cache line
+
+        Args:
+            address: Memory address to modify
+            state: New MESI state to set
+        Returns:
+            True if line exists and state was set, False if line not in cache
+        Raises:
+            ValueError: If invalid state transition attempted
+        """
+        addr_fields = self.__decompose_address(address)
+
+        if self.sets[addr_fields.index] is None:
+            return False
+
+        cache_set = self.sets[addr_fields.index]
+        way_index = cache_set.find_way_by_tag(addr_fields.tag)
+
+        if way_index is not None:
+            try:
+                cache_set.mesi_state[way_index] = state
+                return True
+            except ValueError as e:
+                raise ValueError(
+                    f"Invalid state transition for address {hex(address)}: {e}"
+                )
+        return False
+
+    def clear_cache(self):
+        """Clear all cache contents and reset statistics"""
+        self.sets = [None] * self.num_sets
+        self.allocated_sets = 0
+
```

Comment:
I'm wondering if this is the behavior we want when we clear the cache (deallocate all the sets) or do we just want the contents zeroed out (or set to none)? Not sure that it matters necessarily for the purposes of the simulation, but it is two different things.

Reply:
This frees memory reference to the objects which in theory should be picked up by Python's garbage collection. Its much easier to dereference everything than to go through every line and clear things to zero.

Reply:
Yeah, I get that. I'm just wondering what behavior Mark is looking for (or does he care). Obviously it doesn't affect the functionality of the simulation.